File: csharp/Fasta.cs
using System.Text;

public class Fasta : Benchmark
{
    public long _n;
    private StringBuilder _resultBuilder;

    public Fasta()
    {
        _resultBuilder = new StringBuilder();
        _n = ConfigVal("n");
    }

    public override void Prepare() { }

    private char SelectRandom((char, double)[] genelist)
    {
        double r = Helper.NextFloat();
        if (r < genelist[0].Item2) return genelist[0].Item1;

        int lo = 0;
        int hi = genelist.Length - 1;

        while (hi > lo + 1)
        {
            int i = (hi + lo) / 2;
            if (r < genelist[i].Item2) hi = i;
            else lo = i;
        }
        return genelist[hi].Item1;
    }

    private const int LINE_LENGTH = 60;

    private void MakeRandomFasta(string id, string desc, (char, double)[] genelist, int n)
    {
        _resultBuilder.Append($">{id} {desc}\n");

        int todo = n;
        char[] buffer = new char[LINE_LENGTH];

        while (todo > 0)
        {
            int m = todo < LINE_LENGTH ? todo : LINE_LENGTH;

            for (int i = 0; i < m; i++) buffer[i] = SelectRandom(genelist);

            _resultBuilder.Append(buffer, 0, m);
            _resultBuilder.Append('\n');
            todo -= LINE_LENGTH;
        }
    }

    private void MakeRepeatFasta(string id, string desc, string s, int n)
    {
        _resultBuilder.Append($">{id} {desc}\n");

        int todo = n;
        int k = 0;
        int kn = s.Length;

        while (todo > 0)
        {
            int m = todo < LINE_LENGTH ? todo : LINE_LENGTH;

            while (m >= kn - k)
            {
                _resultBuilder.Append(s, k, kn - k);
                m -= kn - k;
                k = 0;
            }

            if (m > 0)
            {
                _resultBuilder.Append(s, k, m);
                k += m;
            }

            _resultBuilder.Append('\n');
            todo -= LINE_LENGTH;
        }
    }

    private static readonly (char, double)[] IUB = new (char, double)[]
    {
        ('a', 0.27), ('c', 0.39), ('g', 0.51), ('t', 0.78),
        ('B', 0.8), ('D', 0.8200000000000001), ('H', 0.8400000000000001),
        ('K', 0.8600000000000001), ('M', 0.8800000000000001),
        ('N', 0.9000000000000001), ('R', 0.9200000000000002),
        ('S', 0.9400000000000002), ('V', 0.9600000000000002),
        ('W', 0.9800000000000002), ('Y', 1.0000000000000002)
    };

    private static readonly (char, double)[] HOMO = new (char, double)[]
    {
        ('a', 0.302954942668), ('c', 0.5009432431601),
        ('g', 0.6984905497992), ('t', 1.0)
    };

    private const string ALU = "GGCCGGGCGCGGTGGCTCACGCCTGTAATCCCAGCACTTTGGGAGGCCGAGGCGGGCGGATCACCTGAGGTCAGGAGTTCGAGACCAGCCTGGCCAACATGGTGAAACCCCGTCTCTACTAAAAATACAAAAATTAGCCGGGCGTGGTGGCGCGCGCCTGTAATCCCAGCTACTCGGGAGGCTGAGGCAGGAGAATCGCTTGAACCCGGGAGGCGGAGGTTGCAGTGAGCCGAGATCGCGCCACTGCACTCCAGCCTGGGCGACAGAGCGAGACTCCGTCTCAAAAA";

    public override void Run(long IterationId)
    {
        MakeRepeatFasta("ONE", "Homo sapiens alu", ALU, (int)(_n * 2));
        MakeRandomFasta("TWO", "IUB ambiguity codes", IUB, (int)(_n * 3));
        MakeRandomFasta("THREE", "Homo sapiens frequency", HOMO, (int)(_n * 5));
    }

    public string GetResult() => _resultBuilder.ToString();
    public override uint Checksum => Helper.Checksum(_resultBuilder.ToString());
}

File: zig/src/fasta.zig
const std = @import("std");
const Benchmark = @import("benchmark.zig").Benchmark;
const Helper = @import("helper.zig").Helper;

pub const Fasta = struct {
    allocator: std.mem.Allocator,
    helper: *Helper,
    n: i64,
    result_str: std.ArrayListUnmanaged(u8),
    result_val: u32,

    const Gene = struct {
        c: u8,
        prob: f64,
    };

    const LINE_LENGTH: usize = 60;
    const IUB = [_]Gene{
        .{ .c = 'a', .prob = 0.27 },
        .{ .c = 'c', .prob = 0.39 },
        .{ .c = 'g', .prob = 0.51 },
        .{ .c = 't', .prob = 0.78 },
        .{ .c = 'B', .prob = 0.8 },
        .{ .c = 'D', .prob = 0.8200000000000001 },
        .{ .c = 'H', .prob = 0.8400000000000001 },
        .{ .c = 'K', .prob = 0.8600000000000001 },
        .{ .c = 'M', .prob = 0.8800000000000001 },
        .{ .c = 'N', .prob = 0.9000000000000001 },
        .{ .c = 'R', .prob = 0.9200000000000002 },
        .{ .c = 'S', .prob = 0.9400000000000002 },
        .{ .c = 'V', .prob = 0.9600000000000002 },
        .{ .c = 'W', .prob = 0.9800000000000002 },
        .{ .c = 'Y', .prob = 1.0000000000000002 },
    };

    const HOMO = [_]Gene{
        .{ .c = 'a', .prob = 0.302954942668 },
        .{ .c = 'c', .prob = 0.5009432431601 },
        .{ .c = 'g', .prob = 0.6984905497992 },
        .{ .c = 't', .prob = 1.0 },
    };

    const ALU = "GGCCGGGCGCGGTGGCTCACGCCTGTAATCCCAGCACTTTGGGAGGCCGAGGCGGGCGGATCACCTGAGGTCAGGAGTTCGAGACCAGCCTGGCCAACATGGTGAAACCCCGTCTCTACTAAAAATACAAAAATTAGCCGGGCGTGGTGGCGCGCGCCTGTAATCCCAGCTACTCGGGAGGCTGAGGCAGGAGAATCGCTTGAACCCGGGAGGCGGAGGTTGCAGTGAGCCGAGATCGCGCCACTGCACTCCAGCCTGGGCGACAGAGCGAGACTCCGTCTCAAAAA";

    const vtable = Benchmark.VTable{
        .run = runImpl,
        .checksum = resultImpl,
        .deinit = deinitImpl,
    };

    pub fn init(allocator: std.mem.Allocator, helper: *Helper) !*Fasta {
        const n = helper.config_i64("Fasta", "n");

        const self = try allocator.create(Fasta);
        errdefer allocator.destroy(self);

        self.* = Fasta{
            .allocator = allocator,
            .helper = helper,
            .n = n,
            .result_str = .{},
            .result_val = 0,
        };

        return self;
    }

    pub fn deinit(self: *Fasta) void {
        self.result_str.deinit(self.allocator);
        self.allocator.destroy(self);
    }

    pub fn asBenchmark(self: *Fasta) Benchmark {
        return Benchmark.init(self, &vtable, self.helper, "Fasta");
    }

    pub fn getResult(self: *const Fasta) []const u8 {
        return self.result_str.items;
    }

    fn selectRandom(self: *Fasta, genelist: []const Gene) u8 {
        const r = self.helper.nextFloat(1.0);

        if (r < genelist[0].prob) return genelist[0].c;

        var lo: usize = 0;
        var hi: usize = genelist.len - 1;

        while (hi > lo + 1) {
            const i = (hi + lo) / 2;
            if (r < genelist[i].prob) {
                hi = i;
            } else {
                lo = i;
            }
        }

        return genelist[hi].c;
    }

    fn makeRandomFasta(self: *Fasta, id: []const u8, desc: []const u8, genelist: []const Gene, n_iter: i32) void {
        self.result_str.appendSlice(self.allocator, ">") catch return;
        self.result_str.appendSlice(self.allocator, id) catch return;
        self.result_str.appendSlice(self.allocator, " ") catch return;
        self.result_str.appendSlice(self.allocator, desc) catch return;
        self.result_str.appendSlice(self.allocator, "\n") catch return;

        var todo: i32 = n_iter;

        while (todo > 0) {
            const m_val = if (todo < LINE_LENGTH) todo else @as(i32, @intCast(LINE_LENGTH));
            const m = @as(usize, @intCast(m_val));

            var buffer: [LINE_LENGTH]u8 = undefined;

            for (0..m) |i| {
                buffer[i] = self.selectRandom(genelist);
            }

            self.result_str.appendSlice(self.allocator, buffer[0..m]) catch return;
            self.result_str.appendSlice(self.allocator, "\n") catch return;

            todo -= @as(i32, @intCast(LINE_LENGTH));
        }
    }

    fn makeRepeatFasta(self: *Fasta, id: []const u8, desc: []const u8, s: []const u8, n_iter: i32) void {
        self.result_str.appendSlice(self.allocator, ">") catch return;
        self.result_str.appendSlice(self.allocator, id) catch return;
        self.result_str.appendSlice(self.allocator, " ") catch return;
        self.result_str.appendSlice(self.allocator, desc) catch return;
        self.result_str.appendSlice(self.allocator, "\n") catch return;

        var todo: i32 = n_iter;
        var k: usize = 0;
        const kn = s.len;

        while (todo > 0) {
            const m_val = if (todo < LINE_LENGTH) todo else @as(i32, @intCast(LINE_LENGTH));
            var m = @as(usize, @intCast(m_val));

            while (m >= kn - k) {
                self.result_str.appendSlice(self.allocator, s[k..]) catch return;
                m -= (kn - k);
                k = 0;
            }

            self.result_str.appendSlice(self.allocator, s[k .. k + m]) catch return;
            self.result_str.appendSlice(self.allocator, "\n") catch return;

            k += m;
            todo -= @as(i32, @intCast(LINE_LENGTH));
        }
    }

    fn runImpl(ptr: *anyopaque, iteration_id: i64) void {
        _ = iteration_id;
        const self: *Fasta = @ptrCast(@alignCast(ptr));

        const n = @as(i32, @intCast(self.n));
        self.makeRepeatFasta("ONE", "Homo sapiens alu", ALU, n * 2);
        self.makeRandomFasta("TWO", "IUB ambiguity codes", &IUB, n * 3);
        self.makeRandomFasta("THREE", "Homo sapiens frequency", &HOMO, n * 5);
    }

    fn resultImpl(ptr: *anyopaque) u32 {
        const self: *Fasta = @ptrCast(@alignCast(ptr));
        return self.helper.checksumString(self.result_str.items);
    }

    fn deinitImpl(ptr: *anyopaque) void {
        const self: *Fasta = @ptrCast(@alignCast(ptr));
        self.deinit();
    }
};
